### PR TITLE
Add syntax-propertize-function to improve `|||` matching

### LINKED ISSUE #7: Multiline string literals not matched correctly
Some examples:

```
{
  // This gets highlighted as a multiline string, but it should not be:
  foo1: |
    bar
  |;

  // This too:
  foo2: | bar |;

  // This too:
  foo3: = |||||
    bar
  |||||;

  // This one gets interrupted at the | in the middle, and the rest of
  // the file gets highlighted as a string until another | shows up:
  foo4: |||
    bar | this text should still be inside the triple-pipe string
  |||;

  // this line should be a comment but it's highlighted as a string?

  // if I add a single | it ends the incorrect string match

  // now it's back to normal and this is a comment again
}
```

here's what that looks like in a screenshot:

screenshot from 2018-08-27 16-38-30

### PR DESCRIPTION
**Pull request by anguslees**
## Description

Add jsonnet-specific `syntax-propertize-function` hook with custom logic for setting string syntax property on multiline `|||...|||` blocks.

Should correctly deal with complicated cases like:

```
local value = |||
  foo
  |||
    surprise!
  |||
|||;
```

In addition, use the new syntax property to make `jsonnet--find-current-multiline-string` more efficient.

## Motivation and Context

Fixes #7

## How Has This Been Tested?

Tested on:

```
/*  Uncomment to test
local foo = ||
 this is not a string
||;

local foo = |
 nor is this
|;
*/

local foo = |||
  this _is_ a string
|||;


local foo = |||
|||;

local foo = |||

|||;

local foo = |||

  previous line is empty. this is the real prefix
|||;

{
  local foo = |||
    foo
    |||
    surprise!
    |||
  |||;
}

/*
 this is not a string delimiter
 |||
*/

local foo = "||| also not a delimiter";
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project (see https://github.com/bbatsov/emacs-lisp-style-guide).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.

I have never written font-lock/syntax code before, so suggestions/criticism welcome.  I just fumbled through the emacs docs, prompted by the sh-mode here-doc precedent.

### CODE REVIEW

File: jsonnet-mode.el [hunk 1]
```diff
@@ -237,7 +287,10 @@ If not inside of a multiline string, return nil."
                                                    nil                        ;; syntax-alist
                                                    nil                        ;; syntax-begin
                                                    ))
-  (set (make-local-variable 'indent-line-function) 'jsonnet-indent))
+  (set (make-local-variable 'indent-line-function) 'jsonnet-indent)
+  (setq-local syntax-propertize-function #'jsonnet-syntax-propertize-function)
```
[the remark this answers is not in the document]
Comment (anguslees) [reply jsonnet-mode.el]: Oops, that will teach me to try to follow a style guide at the last moment.  Fixed.

**Comment by benley** [comment]
Ping? This seems like it is ready to be merged.

**Comment by sparkprime** [comment]
:+1:

**Comment by legonigel** [comment]
@mgyucht Could this get merged?

**Comment by mgyucht** [comment]
TBH I don't use emacs anymore, so I'm not maintaining this, but I'm glad to merge it. Any chance anyone here wants to take ownership of this?

**Comment by tminor** [comment]
Hi @mgyucht. Is it correct to interpret "take ownership" as "maintain this project"? I'd be willing to do it if so!

**Comment by mgyucht** [comment]
@tminor If you want, I'd be glad to transfer this to you. I think you need to delete your fork before I can change it to you.

**Comment by tminor** [comment]
Done and thanks for the quick reply, @mgyucht!

**Pull request merged**